### PR TITLE
:gift: Author notifications

### PR DESCRIPTION
**Pull request by ShanaLMoore**
Issue: 
- https://github.com/scientist-softserv/palni-palci/issues/1020

TODO:

- [x] make email dynamic, including the date, tenant name, link to dashboard, etc
- [x] Do we want to make an option for the user to not receive emails? If so - we need UI work and possibly migrations. [update - this is now in the SOW and needs to be a separate ticket/work]
- [ ] How often Hyrax::UserStatImporter job run? - Is Rob setting up a job to run the rake task. Task needs to work for all tenants. [ update - not sure this matters for our sake ] 
- [x] specs

image

### CODE REVIEW

File: depositor_email_notification_job.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class DepositorEmailNotificationJob < ApplicationJob
+  non_tenant_job
+
+  after_perform do |job|
+    reenqueue(job.arguments.first)
+  end
+
+  def perform(account)
+    Apartment::Tenant.switch(account.tenant) do
+      stats_period = (Date.today - 1.month).beginning_of_month..(Date.today - 1.month).end_of_month
```
Comment (ShanaLMoore) [inline depositor_email_notification_job.rb:12]: I'm not sure how UserStarts works. It seems like it's possible for user.stats to return multiple UserStats, which we'd need to consider.

```ruby
=>                               
[#<UserStat:0x0000ffff8794a2a0   
  id: 1,                         
  user_id: 1,                    
  date:                          
   Tue, 28 May 2024 00:00:00.000000000 UTC +00:00,                
  file_views: 4,                 
  file_downloads: 3,             
  created_at:
   Wed, 29 May 2024 18:14:22.418822000 UTC +00:00,
  updated_at:
   Wed, 29 May 2024 18:14:22.418822000 UTC +00:00,
  work_views: nil>]
```

File: app/views/hyku_mailer/depositor_email.html.erb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      padding: 10px 20px;
+      margin: 10px 0;
+      background-color: #0056b3;
+      color: white;
+      text-decoration: none;
+      border-radius: 5px;
+    }
+  </style>
+</head>
+<body>
```
Comment (ShanaLMoore) [inline app/views/hyku_mailer/depositor_email.html.erb:20]: This needs to be dynamic. Currently hardcoded placeholder data is in play.

File: spec/mailers/previews/hyku_mailer_preview.rb [hunk 3]
```diff
@@ -21,4 +21,25 @@ def summary_email
 
     HykuMailer.new.summary_email(user, messages, account)
   end
+
```
Comment (ShanaLMoore) [inline spec/mailers/previews/hyku_mailer_preview.rb:24]: to view, create an acme tenant and visit http://acme.pals.test/rails/mailers/hyku_mailer/depositor_email

File: depositor_email_notification_job.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# frozen_literal: true
+
+# TODO
+# How do we collect all of the counts for the past month only?
+# Do we expect user.stats to return multiple UserStat objects. How does this work?
```
Comment (ShanaLMoore) [inline depositor_email_notification_job.rb:5]: Per Rob, we can expect multi UserStats, which get created for each day. We'd need to find them all by date for the past month and sum the totals to report the new ones. 

`>= '2024-04-01' AND < '2024-05-01'` => find everything for the month of April. 

```ruby
[#<UserStat:0x0000ffff85feb188
  id: 1,
  user_id: 1,
  date: Thu, 23 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.429000000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.429000000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feb048
  id: 2,
  user_id: 1,
  date: Fri, 24 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.432511000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.432511000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feaf08
  id: 3,
  user_id: 1,
  date: Sat, 25 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.435427000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.435427000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feadc8
  id: 4,
  user_id: 1,
  date: Sun, 26 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.437734000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.437734000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feac88
  id: 5,
  user_id: 1,
  date: Mon, 27 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.439853000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.439853000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feab48
  id: 6,
  user_id: 1,
  date: Tue, 28 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.441642000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.441642000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feaa08
  id: 7,
  user_id: 1,
  date: Wed, 29 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.443533000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.443533000 UTC +00:00,
  work_views: 0>]
```

File: depositor_email_notification_job.rb (hunk 1 above)
Comment (ShanaLMoore) [reply depositor_email_notification_job.rb:18]: Per Rob, we can expect multi UserStats, which get created for each day. We'd need to find them all by date for the past month and sum the totals to report the new ones. 

`>= '2024-04-01' AND < '2024-05-01'` => find everything for the month of April. 

```ruby
[#<UserStat:0x0000ffff85feb188
  id: 1,
  user_id: 1,
  date: Thu, 23 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.429000000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.429000000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feb048
  id: 2,
  user_id: 1,
  date: Fri, 24 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.432511000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.432511000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feaf08
  id: 3,
  user_id: 1,
  date: Sat, 25 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.435427000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.435427000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feadc8
  id: 4,
  user_id: 1,
  date: Sun, 26 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.437734000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.437734000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feac88
  id: 5,
  user_id: 1,
  date: Mon, 27 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.439853000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.439853000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feab48
  id: 6,
  user_id: 1,
  date: Tue, 28 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.441642000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.441642000 UTC +00:00,
  work_views: 0>,
 #<UserStat:0x0000ffff85feaa08
  id: 7,
  user_id: 1,
  date: Wed, 29 May 2024 00:00:00.000000000 UTC +00:00,
  file_views: 0,
  file_downloads: 0,
  created_at: Thu, 30 May 2024 23:32:22.443533000 UTC +00:00,
  updated_at: Thu, 30 May 2024 23:32:22.443533000 UTC +00:00,
  work_views: 0>]
```

**Comment by ShanaLMoore** [comment]
I can't approve this since I opened it, but it looks good! Thanks so much for completing this.

**Review by kirkkwang — DISMISSED** [review]

**Review by kirkkwang — APPROVED** [review]

File: app/views/hyku_mailer/depositor_email.html.erb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  <div class="email-container">
+    <p>Dear Author,</p>
+    <p>You had <%= pluralize(@statistics[:new_file_downloads], 'download') %> in <%= Date.today.prev_month.strftime("%B %Y") %> across your <%= pluralize(@statistics[:new_work_views], 'work') %> in <%= @application_name %>. Your current readership:</p>
+    <ul>
+      <li><%= pluralize(@statistics[:total_file_downloads], 'Total File Download') %></li>
+      <li><%= pluralize(@statistics[:total_work_views], 'Total Work View') %></li>
+    </ul>
+    <a href="<%= @url %>" class="button">VISIT MY DASHBOARD</a>
+    <div class="footer">
+      <p>These monthly reports are provided to you by bepress on behalf of <%= @application_name %>. For questions, comments, or to add more content and increase your readership and visibility as an author, please contact your repository administrator.</p>
```
Comment (ShanaLMoore) [inline app/views/hyku_mailer/depositor_email.html.erb:30]: what is bepress

Comment (kirkkwang) [reply app/views/hyku_mailer/depositor_email.html.erb:30]: Looks like that was in the original template and was simply copy and pasted in.  https://bepress.com/

**Review by kirkkwang — APPROVED** [review]

**Pull request merged**